engine/source_sym_engine.ml

```diff
@@ -61,122 +20,102 @@ let print_result stree =
+      | Cons -> bprintf buf "Cons"
+    in
     match tree with
-    | Leaf (pilist, expr) ->
+    | Leaf expr ->
       bprintf buf
-        "Leaf='%a':\
-         %t"
+        "Leaf='%a'"
         bprint_source_expr expr
-        (break (ntabs+1)) ;
-      bprintf buf "constraints {%a}" (bprint_list ~sep bprint_pi) pilist
-    | Node (pi, cst_list) ->
-      bprintf buf
-         "Node: constraints {%a}\
-          %t%a"
-         bprint_pi pi
-         (break (ntabs+1))
-         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
-    | Conjunction (cst_list) ->
-      bprintf buf
-         "Conjunction:%t%a"
-         (break (ntabs+1))
-         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
-    | Root cst_list ->
-       bprintf buf
-         "Root=\
-          %t%a"
-         (break (ntabs+1))
-         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Node (k_cst_list, cst_opt) ->
+      bprintf buf "Node:{\
+                   %a \
+                   %t} Fallback: %a"
+        (bprint_list ~sep:sep
+           (fun buf (k,cst) -> bprintf buf "%t%a -> %t%a"
+               sep
+               bprint_constructor k
+               (break (ntabs+2))
+               (bprint_tree (ntabs+1)) cst))
+        k_cst_list
+        (break ntabs)
+        (fun buf -> begin match cst_opt with 
+             | Some cst -> fun _ -> bprint_tree (ntabs+1) buf cst
+             | None -> fun _ -> bprintf buf "None" 
+           end) buf
   in
   let buf = Buffer.create 42 in
   bprint_tree 0 buf stree;
   BatIO.write_line BatIO.stdout (Buffer.contents buf)
 
-let sym_exec (source: source_program) =
-  let shares_kst ~wildcard_to_none k ((pattern, expr):clause): clause option =
-    match pattern with
-    | Wildcard when wildcard_to_none = false-> Some (Wildcard, expr)
-    | Constructor (k', plist) when k' = k ->
-      if (List.length plist > 1) then
-        Some (Constructor (Tuple, plist), expr)
-      else
-        Some (plist |> List.hd, expr)
-    | _ -> None
-  in
-  let negate_piop = function
-    | EqInt i -> NeqInt i
-    | NeqInt i -> EqInt i
-    | IsK s -> NotK s
-    | NotK s -> IsK s
-    | IsTrue -> IsFalse
-    | IsFalse -> IsTrue
-    | IsTuple -> NotTuple
-    | NotTuple -> IsTuple
-    | IsNil -> NotNil
-    | NotNil -> IsNil
-    | IsCons -> NotCons
-    | NotCons -> IsCons
-    | EqString s -> NeqString s
-    | NeqString s -> EqString s
-  in
-  let piop_of_constructor = function
-    | Variant s -> IsK s
-    | Int i -> EqInt i
-    | Bool b -> if b then IsTrue else IsFalse
-    | String s -> EqString s
-    | Tuple -> IsTuple
-    | Nil -> IsNil
-    | Cons -> IsCons
+type row = pattern list * source_expr
+
+let sym_exec source =
+  let kst_rows_assoc rows : (constructor * row list) list * row option =
+    let hashtbl = Hashtbl.create 42
+    in
+    let with_wildcard = ref []
+    in
+    let relevant_wildcard_opt = function
+      | [] -> None
+      | wildcards -> Some (wildcards |> List.rev |> List.hd)
```

Comment:
Be careful, I don't think that discarding wildcard rows except the first one is valid for pattern matrices, because you don't know that the *other* columns are subsumed by the first row. Consider for example

```
(_ true) -> e1
(_ false) -> e2
```